Implement private methods

Akirathan opened the pull request:
Closes #8834

### Pull Request Description

Add support for private methods. Most of the changes are in parser and compiler. The runtime checking of private functions was already present since #9692

### Important Notes

- Only top-level methods can be declared `private`.
- private method cannot be called from different project
- private method cannot be accessed from polyglot code (private method does not exist for polyglot code)

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      [TypeScript](https://github.com/enso-org/enso/blob/develop/docs/style-guide/typescript.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- [x] Unit tests have been written where possible.

Commit:
Move TestBase into a new project test-utils

Commit:
Parser supports private methods

Commit:
Add isPrivate flag to various method IR classes

Commit:
Add parser and compiler tests

Commit:
Update docs

Commit:
IrToTruffle sets FuncSchema as project private

Commit:
Rename PrivateAccessTest

Commit:
Add more compiler tests

Commit:
private is a keyword in TextMate grammar

Commit:
Add basic tests to Private_Spec.enso

Commit:
Fix private method recognition in IrToTruffle

Commit:
Add another parser test

Commit:
fmt

Commit:
Revert "Move TestBase into a new project test-utils"

This reverts commit 4b18d5da

Commit:
private methods are inaccessible to polyglot code

Commit:
Merge branch 'refs/heads/develop' into wip/akirathan/8834-private-methods

# Conflicts:
#	engine/runtime-parser/src/main/java/org/enso/compiler/core/TreeToIr.java
#	engine/runtime-parser/src/test/java/org/enso/compiler/core/EnsoParserTest.java

Commit:
Remove unused cmdline options from runtime-parser/Test

Commit:
fmt

Akirathan commented on engine/runtime-parser/src/main/java/org/enso/compiler/core/TreeToIr.java:
A lot of modifications in this source are formatting. I have noticed that the indentation in this source is broken and that our formatter does not notice that, so I did a force formatting.

Commit:
Update changelog

kazcw reviewed (approved):
Parser changes LGTM

Commit:
Private methods are not in Suggestion DB

radeusgd commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/ErrorCompilerTest.java:
```diff
@@ -465,6 +465,46 @@ public void illegalForeignBody4() throws Exception {
+            private priv_nested_method x = x
+        """);
+    assertSingleSyntaxError(
+        ir, Syntax.UnexpectedExpression$.MODULE$, "Unexpected expression", 13, 45);
```
Could we have clearer errors than that? This is really not helpful for end-users.

It would be much better if it could say something like "inner methods cannot be `private`"

Akirathan replied:
In this particular case, it is not that simple to have a better error message. In many other places (see for example tests in the parser), the error message is much more descriptive. This is just handled by some fallback method in `TreeToIr` and it is not simple to change that.

radeusgd commented on tools/enso4igv/src/main/resources/org/enso/tools/enso4igv/enso.tmLanguage.json:
```diff
@@ -21,7 +21,7 @@
 			"patterns": [{
 					"name": "keyword.other",
-					"match": "\\b(self|type)\\b"
+					"match": "\\b(self|type|private)\\b"
```
👍

radeusgd reviewed (approved):
Looks good overall. It would be great if some error messages were clearer, if not in this PR, I'd at least consider creating a ticket for this.

I'm also curious as I haven't seen this tested - what about extension methods? Is the `private` keyword allowed for them? I'd like to see a test, either positive or negative for it.

Moreover, you say that `private` is not allowed on conversions - but I don't see _any_ test that verifies what happens if I put it there? Will I get an error? What will the error say? Or will the private keyword just be ignored?

JaroslavTulach commented on docs/semantics/encapsulation.md:
```diff
@@ -74,10 +74,14 @@ public and private constructors in a single type is a compilation error. A type
 private method is accessed only from within the same project, a runtime check
 must be performed, as this cannot be checked during the compilation.
 
+Conversion and foreign methods cannot be specified as private.
```
There probably is no (technical) reason why a `foreign` method couldn't be `private`, I think.

Akirathan replied:
Correct. Other than probably non-trivial changes in the parser once again. The following snippet:
```
private foreign js js_func = """
    return 42; 
```
parses to:
```
(Private (Function (Ident foreign) #((() (Ident js) () ()) (() (Ident js_func) () ())) () "=" (TextLiteral #((Section "return 42;")))))
```
and I don't know (and don't want to deal with it in this PR) how to convince the parser to parse it differently.

Let's skip the private declaration for foreign methods for now. If there is a use-case for them in the future, let's create another ticket for that.

Akirathan replied:
Created a follow-up https://github.com/enso-org/enso/issues/10137 for this. Let's not solve private foreign functions in this PR.

JaroslavTulach commented on engine/runtime-integration-tests/src/test/scala/org/enso/compiler/test/context/SuggestionBuilderTest.scala:
```diff
@@ -3143,6 +3143,84 @@ class SuggestionBuilderTest extends AnyWordSpecLike with Matchers {
+    "build type with private methods" in {
+      val code =
+        """type T
+          |    private priv_meth self = 42
```
Suggestion builder test! Great.

JaroslavTulach reviewed (approved):
Good change.

Commit:
Add test for "Conversion methods cannot be specified as private"

Commit:
Add test for "Extension method can be specified as private"

Akirathan commented:
> Moreover, you say that private is not allowed on conversions - but I don't see any test that verifies what happens if I put it there? Will I get an error? What will the error say? Or will the private keyword just be ignored?

@radeusgd Valid point. I have added a test for that in 68225cf5b80993856eb65931914ce3cfc2758afb. By the way, the reason why conversion methods *cannot* be specified as private is because of the different, more complicated, execution mechanism for the conversions. Let's not complicate things. I don't know how I would implement the conversion dispatch if one conversion from a different project is private, but another one from yet different project is not. If there will be a use case for private conversions, let's create a new issue for that.

> I'm also curious as I haven't seen this tested - what about extension methods? Is the private keyword allowed for them? I'd like to see a test, either positive or negative for it.

Extension method are basically normal methods. They can be specified as private as well. I added a test for that in 1029ab299181b43042f24e5804be459dae4772e4.

radeusgd commented:
> By the way, the reason why conversion methods _cannot_ be specified as private is because of the different, more complicated, execution mechanism for the conversions. Let's not complicate things. I don't know how I would implement the conversion dispatch if one conversion from a different project is private, but another one from yet different project is not. If there will be a use case for private conversions, let's create a new issue for that.

Oh of course I did not mean to suggest we want private conversion methods! I think that would be very messy. I 💯 agree with you that it makes sense to not allow marking them private. My comment was _only_ about a need to have a test that demonstrates them working in one way or another (e.g. displaying an error telling that this is not allowed).

radeusgd commented on engine/runtime-integration-tests/src/test/java/org/enso/compiler/ErrorCompilerTest.java:
```diff
@@ -505,6 +505,18 @@ public void illegalPrivateTypeDeclaration() throws Exception {
+        private My_Type.from (other:Other_Type) = 42
+        """);
+    assertSingleSyntaxError(
+        ir, Syntax.UnexpectedExpression$.MODULE$, "Unexpected expression", 29, 73);
```
Is there any way that we could get more descriptive error messages than `"Unexpected expression"`?

Commit:
Merge branch 'refs/heads/develop' into wip/akirathan/8834-private-methods

Commit:
Conversion methods cannot be specified as private

Commit:
Remove test for private foreign function testing.

Will be solved in https://github.com/enso-org/enso/issues/10137

Commit:
Merge branch 'refs/heads/develop' into wip/akirathan/8834-private-methods

# Conflicts:
#	CHANGELOG.md
#	engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/privateaccess/PrivateConstructorAccessTest.java

Commit:
Fix test after merge of ContextUtils

Akirathan commented:
Parsing of private foreign functions will be a follow-up tracked in #10137. Let's merge this PR ASAP.

Commit:
Fix class name after merge
